packages/solana/src/client.ts

```diff
@@ -0,0 +1,482 @@
+import type UniversalProvider from '@walletconnect/universal-provider'
+import type {
+  CaipNetworkId,
+  ConnectionControllerClient,
+  LibraryOptions,
+  NetworkControllerClient,
+  Token,
+  ScaffoldOptions,
+  Connector,
+  CaipAddress,
+  CaipNetwork
+} from '@web3modal/scaffold'
+import type {
+  ProviderType,
+  Chain,
+  Provider,
+  SolStoreUtilState
+} from '@web3modal/scaffold-utils/solana'
+import type { Web3ModalSIWEClient } from '@web3modal/siwe'
+
+export interface Web3ModalClientOptions extends Omit<LibraryOptions, 'defaultChain' | 'tokens'> {
+  solanaConfig: ProviderType
+  siweConfig?: Web3ModalSIWEClient
+  chains: Chain[]
+  defaultChain?: Chain
+  chainImages?: Record<number | string, string>
+  connectorImages?: Record<string, string>
+  tokens?: Record<number, Token>
+}
+
+export type Web3ModalOptions = Omit<Web3ModalClientOptions, '_sdkVersion'>
+
+// -- Client --------------------------------------------------------------------
+export class Web3Modal extends Web3ModalScaffold {
+  private hasSyncedConnectedAccount = false
+
+  private WalletConnectConnector: WalletConnectConnector
+  private walletAdapters: Record<AdapterKey, BaseWalletAdapter>
+
+  private chains: Chain[]
+
+  public constructor(options: Web3ModalClientOptions) {
+    const { solanaConfig, chains, tokens, _sdkVersion, chainImages, ...w3mOptions } = options
+
+    if (!solanaConfig) {
+      throw new Error('web3modal:constructor - solanaConfig is undefined')
+    }
+
+    if (!w3mOptions.projectId) {
+      throw new Error('web3modal:constructor - projectId is undefined')
+    }
+
+    const networkControllerClient: NetworkControllerClient = {
+      switchCaipNetwork: async caipNetwork => {
+        if (caipNetwork) {
+          try {
+            await this.switchNetwork(caipNetwork)
+          } catch (error) {
+            SolStoreUtil.setError(error)
+          }
+        }
+      },
+
+      getApprovedCaipNetworksData: async () =>
+        new Promise(resolve => {
+          const result = {
+            approvedCaipNetworkIds: undefined,
+            supportsAllNetworks: true
+          }
+
+          resolve(result)
+        })
+    }
+
+    const connectionControllerClient: ConnectionControllerClient = {
+      connectWalletConnect: async onUri => {
+        const WalletConnectProvider = await this.WalletConnectConnector.getProvider()
+        if (!WalletConnectProvider) {
+          throw new Error('connectionControllerClient:getWalletConnectUri - provider is undefined')
+        }
+
+        WalletConnectProvider.on('display_uri', (uri: string) => {
+          onUri(uri)
+        })
+        const address = await this.WalletConnectConnector.connect()
+        this.setWalletConnectProvider(address)
+      },
+
+      connectExternal: async ({ id }) => {
+        const adapter =
+          id.toLocaleLowerCase() === 'Trust'
+            ? 'trustWallet'
+            : (id.toLocaleLowerCase() as AdapterKey)
```

Comment:
do we have any control over this data? If so, could we add a `name` property to avoid manually mapping special cases?